reference-implementation/lib/abstract-ops/readable-streams.js

```diff
@@ -1570,6 +1595,10 @@ function ReadableByteStreamControllerRespondInReadableState(controller, bytesWri
     return;
   }
 
+  if (pullIntoDescriptor.readFully && pullIntoDescriptor.bytesFilled < pullIntoDescriptor.byteLength) {
+    return;
```

Comment:
Maybe add a comment here (eventually a spec note) explaining that, IIUC, this is where the magic happens. If we're trying to read fully and haven't filled things up yet, just return, leaving the modified pull-into descriptor in the queue. The next time around we'll pick up right back here, until we're full.

Reply:
There are two places where the magic happens:
* Here, in `ReadableByteStreamControllerRespondInReadableState`, we make sure not to shift the first pull-into descriptor if a call to `respond()` or `respondWithNewView()` did not fill the entire view.
* [Over in `ReadableByteStreamControllerFillPullIntoDescriptorFromQueue`](https://github.com/whatwg/streams/commit/383c9e543a63938a2c2497a627b0a216d89e7250#diff-db46b8b33e26be52c078fa1540a3516fca7f75af6effb18e9b152efbd114b4bfL1350-R1370), we make sure to indicate that the pull-into descriptor is "not ready" if the total queue size is smaller than the view's length. This is used by the callers to decide if the pull-into descriptor should be shifted out or not. This handles the case where there were already some chunks queued up before `readFully()` was called, or if the underlying source uses `enqueue()` to (partially) fill up the pull-into descriptor.

I'll try to explain all of this in a code comment. 🙂